File: tests/typesafety/python/alias_compound.py
# expect: passes
# title: RetryPolicy and DEFAULT_BACKOFF give the type and the number names
# improves: alias_repeated
from collections.abc import Callable
from typing import Final

type Milliseconds = int
type RetryPolicy = Callable[[int, Exception], Milliseconds]

DEFAULT_BACKOFF: Final[Milliseconds] = 30_000


def fixed_backoff(_attempt: int, _error: Exception) -> Milliseconds:
    return DEFAULT_BACKOFF


def doubling_backoff(attempt: int, _error: Exception) -> Milliseconds:
    return DEFAULT_BACKOFF * (1 << attempt)


def run_with_retries(policy: RetryPolicy) -> Milliseconds:
    return policy(1, ValueError("transient"))

File: tests/typesafety/typescript/alias_compound.ts
// expect: passes

type Milliseconds = number;
type RetryPolicy = (attempt: number, error: Error) => Milliseconds;

const DEFAULT_BACKOFF: Milliseconds = 30_000;

export const fixedBackoff: RetryPolicy = () => DEFAULT_BACKOFF;

export const doublingBackoff: RetryPolicy = (attempt) => DEFAULT_BACKOFF * 2 ** attempt;

export function runWithRetries(policy: RetryPolicy): Milliseconds {
  return policy(1, new Error("transient"));
}
